src/crypto/aes.cpp

```diff
@@ -0,0 +1,560 @@
+   slices[5] = a4_01 ^ a5_123;
+   slices[6] = a5_01 ^ a6_123;
+   slices[7] = a6_01 ^ a7_123;
+}
+
+void InvMixColumns(uint16_t* slices) {
+    // b(r,c) = 0e * a(r,c) + 0b * a(r+1,c) + 0d * a(r+2,c) + 09 * a(r+3,c)
+    // b(r,c) = 08 * (a(r,c) + a(r+1,c) + a(r+2,c) + a(r+3,c)) +
+    //          04 * (a(r,c) + a(r+2,c)) +
+    //          02 * (a(r,c) + a(r+1,c)) +
+    //          01 * (a(r+1,c) + a(r+2,c) + a(r+3,c))
+
+   uint16_t a0 = slices[0], a1 = slices[1], a2 = slices[2], a3 = slices[3];
+   uint16_t a4 = slices[4], a5 = slices[5], a6 = slices[6], a7 = slices[7];
+
+   uint16_t a0_01 = a0 ^ ROT(a0, 1), a0_12 = ROT(a0_01, 1), a0_123 = a0_12 ^ ROT(a0, 3), a0_0123 = a0 ^ a0_123, a0_02 = a0_01 ^ a0_12;
+   uint16_t a1_01 = a1 ^ ROT(a1, 1), a1_12 = ROT(a1_01, 1), a1_123 = a1_12 ^ ROT(a1, 3), a1_0123 = a1 ^ a1_123, a1_02 = a1_01 ^ a1_12;
+   uint16_t a2_01 = a2 ^ ROT(a2, 1), a2_12 = ROT(a2_01, 1), a2_123 = a2_12 ^ ROT(a2, 3), a2_0123 = a2 ^ a2_123, a2_02 = a2_01 ^ a2_12;
+   uint16_t a3_01 = a3 ^ ROT(a3, 1), a3_12 = ROT(a3_01, 1), a3_123 = a3_12 ^ ROT(a3, 3), a3_0123 = a3 ^ a3_123, a3_02 = a3_01 ^ a3_12;
+   uint16_t a4_01 = a4 ^ ROT(a4, 1), a4_12 = ROT(a4_01, 1), a4_123 = a4_12 ^ ROT(a4, 3), a4_0123 = a4 ^ a4_123, a4_02 = a4_01 ^ a4_12;
+   uint16_t a5_01 = a5 ^ ROT(a5, 1), a5_12 = ROT(a5_01, 1), a5_123 = a5_12 ^ ROT(a5, 3), a5_0123 = a5 ^ a5_123, a5_02 = a5_01 ^ a5_12;
+   uint16_t a6_01 = a6 ^ ROT(a6, 1), a6_12 = ROT(a6_01, 1), a6_123 = a6_12 ^ ROT(a6, 3), a6_0123 = a6 ^ a6_123, a6_02 = a6_01 ^ a6_12;
+   uint16_t a7_01 = a7 ^ ROT(a7, 1), a7_12 = ROT(a7_01, 1), a7_123 = a7_12 ^ ROT(a7, 3), a7_0123 = a7 ^ a7_123, a7_02 = a7_01 ^ a7_12;
+
+   slices[0] = a0_123 ^ a7_01 ^ a6_02 ^ a5_0123;
+   slices[1] = a1_123 ^ a0_01 ^ a7_12 ^ a6_02 ^ a5_0123 ^ a6_0123;
+   slices[2] = a2_123 ^ a1_01 ^ a0_02 ^ a7_02 ^ a6_0123 ^ a7_0123;
+   slices[3] = a3_123 ^ a2_01 ^ a7_01 ^ a1_02 ^ a6_02 ^ a0_0123 ^ a5_0123 ^ a7_0123;
+   slices[4] = a4_123 ^ a3_01 ^ a7_12 ^ a2_02 ^ a6_02 ^ a1_0123 ^ a5_0123 ^ a6_0123;
+   slices[5] = a5_123 ^ a4_01 ^ a3_02 ^ a7_02 ^ a2_0123 ^ a6_0123 ^ a7_0123;
+   slices[6] = a6_123 ^ a5_01 ^ a4_02 ^ a3_0123 ^ a7_0123;
+   slices[7] = a7_123 ^ a6_01 ^ a5_02 ^ a4_0123;
+}
+
+void AddRoundKey(uint16_t* slices, const AESSlicedMatrix* rk) {
+    for (int b = 0; b < 8; b++) {
+        slices[b] ^= rk->slice[b];
+    }
+}
+
+/** Expand the cipher key into the key schedule. */
+void RijndaelSetup(AESSlicedMatrix* state, const uint8_t* key, int nkeywords, int nrounds)
+{
+    /* The one-byte round constant */
+    uint8_t rcon = 0x01;
+    /* A ring buffer containing the last 8 round key words (4 are consumed in every round) */
+    uint32_t rk[8];
+    /* The round number */
+    int round = 0;
+    /* The number of the word being generated, modulo nkeywords */
+    int pos = 0;
+
+    /* The first nkeywords round key words are just taken from the key directly */
+    for (int i = 0; i < nkeywords; i++) {
```

Comment:
A bounds assertion on nkeywords would be helpful prior to this line (e.g. at the start of the function). The loop requires that i<8 to prevent overflow on the stack, but this is only enforced in the caller.